File: Antidote/FullscreenPicker.m
//
//  FullscreenPicker.m
//  Antidote
//
//  Created by Dmytro Vorobiov on 09/09/15.
//  Copyright (c) 2015 dvor. All rights reserved.
//

#import <Masonry/Masonry.h>
#import "FullscreenPicker.h"
#import "AppearanceManager.h"

static const CGFloat kAnimationDuration = 0.3;
static const CGFloat kToolbarHeight = 44.0;

@interface FullscreenPicker () <UIPickerViewDelegate, UIPickerViewDataSource>

@property (strong, nonatomic) UIButton *blackoutButton;
@property (strong, nonatomic) UIToolbar *toolbar;
@property (strong, nonatomic) UIPickerView *picker;

@property (strong, nonatomic) MASConstraint *pickerBottomConstraint;

@property (strong, nonatomic) NSArray *stringsArray;

@end

@implementation FullscreenPicker

#pragma mark -  Lifecycle

- (instancetype)initWithStrings:(NSArray *)stringsArray selectedIndex:(NSUInteger)index
{
    self = [super init];

    if (! self) {
        return nil;
    }

    self.stringsArray = stringsArray;
    self.backgroundColor = [UIColor clearColor];

    self.blackoutButton = [UIButton new];
    self.blackoutButton.backgroundColor = [UIColor colorWithWhite:1.0 alpha:0.3];
    [self.blackoutButton addTarget:self action:@selector(doneButtonPressed) forControlEvents:UIControlEventTouchUpInside];
    [self addSubview:self.blackoutButton];

    self.toolbar = [UIToolbar new];
    self.toolbar.tintColor = [UIColor whiteColor];
    self.toolbar.barTintColor = [[AppContext sharedContext].appearance loginNavigationBarColor];
    self.toolbar.items = @[
        [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:self action:nil],
        [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                      target:self
                                                      action:@selector(doneButtonPressed)],
    ];
    [self addSubview:self.toolbar];

    self.picker = [UIPickerView new];
    self.picker.backgroundColor = [UIColor whiteColor];
    self.picker.delegate = self;
    self.picker.dataSource = self;
    [self addSubview:self.picker];

    [self.picker selectRow:index inComponent:0 animated:NO];

    [self installConstraints];

    return self;
}

#pragma mark -  Actions

- (void)doneButtonPressed
{
    [self.delegate fullscreenPicker:self willDismissWithSelectedIndex:[self selectedRow]];

    [self hide];
}

#pragma mark -  Public

- (void)showAnimatedInView:(UIView *)view
{
    [view addSubview:self];
    [self makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(view);
    }];

    [self show];
}

#pragma mark -  UIPickerViewDataSource

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    return self.stringsArray.count;
}

#pragma mark -  UIPickerViewDelegate

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component
{
    return self.stringsArray[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    [self.picker reloadAllComponents];
}

#pragma mark -  Private

- (void)installConstraints
{
    [self.blackoutButton makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self);
    }];

    [self.toolbar makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.equalTo(self.picker.top);
        make.height.equalTo(kToolbarHeight);
        make.width.equalTo(self);
    }];

    [self.picker makeConstraints:^(MASConstraintMaker *make) {
        self.pickerBottomConstraint = make.bottom.equalTo(self);
    }];
}

- (void)show
{
    self.blackoutButton.alpha = 0.0;
    self.pickerBottomConstraint.offset(self.picker.frame.size.height);

    [self layoutIfNeeded];

    [UIView animateWithDuration:kAnimationDuration animations:^{
        self.blackoutButton.alpha = 1.0;
        self.pickerBottomConstraint.offset(0.0);

        [self layoutIfNeeded];
    }];
}

- (void)hide
{
    [UIView animateWithDuration:kAnimationDuration animations:^{
        self.blackoutButton.alpha = 0.0;
        self.pickerBottomConstraint.offset(self.picker.frame.size.height);

        [self layoutIfNeeded];

    } completion:^(BOOL f) {
        [self removeFromSuperview];
    }];
}

- (NSInteger)selectedRow
{
    return [self.picker selectedRowInComponent:0];
}

@end

File: Antidote/FullscreenPicker.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import UIKit
import SnapKit

private struct Constants {
    static let AnimationDuration = 0.3
    static let ToolbarHeight: CGFloat = 44.0
}

protocol FullscreenPickerDelegate: class {
    func fullscreenPicker(_ picker: FullscreenPicker, willDismissWithSelectedIndex index: Int)
}

class FullscreenPicker: UIView {
    weak var delegate: FullscreenPickerDelegate?

    fileprivate var theme: Theme

    fileprivate var blackoutButton: UIButton!
    fileprivate var toolbar: UIToolbar!
    fileprivate var picker: UIPickerView!

    fileprivate var pickerBottomConstraint: Constraint!

    fileprivate let stringsArray: [String]

    init(theme: Theme, strings: [String], selectedIndex: Int) {
        self.theme = theme
        self.stringsArray = strings

        super.init(frame: CGRect.zero)

        configureSelf()
        createSubviews()
        installConstraints()

        picker.selectRow(selectedIndex, inComponent: 0, animated: false)
    }

    required convenience init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func showAnimatedInView(_ view: UIView) {
        view.addSubview(self)

        snp.makeConstraints {
            $0.edges.equalTo(view)
        }

        show()
    }
}

// MARK: Actions
extension FullscreenPicker {
    @objc func doneButtonPressed() {
        delegate?.fullscreenPicker(self, willDismissWithSelectedIndex: picker.selectedRow(inComponent: 0))
        hide()
    }
}

extension FullscreenPicker: UIPickerViewDataSource {
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }

    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return stringsArray.count
    }
}

extension FullscreenPicker: UIPickerViewDelegate {
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return stringsArray[row]
    }

    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        picker.reloadAllComponents()
    }
}

private extension FullscreenPicker {
    func configureSelf() {
        backgroundColor = .clear
    }

    func createSubviews() {
        blackoutButton = UIButton()
        blackoutButton.backgroundColor = theme.colorForType(.TranslucentBackground)
        blackoutButton.addTarget(self, action: #selector(FullscreenPicker.doneButtonPressed), for:.touchUpInside)
        blackoutButton.accessibilityElementsHidden = true
        blackoutButton.isAccessibilityElement = false
        addSubview(blackoutButton)

        toolbar = UIToolbar()
        toolbar.tintColor = theme.colorForType(.LoginButtonText)
        toolbar.barTintColor = theme.loginNavigationBarColor
        toolbar.items = [
            UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: self, action: nil),
            UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(FullscreenPicker.doneButtonPressed))
        ]
        addSubview(toolbar)

        picker = UIPickerView()
        // Picker is always white, despite choosen theme
        picker.backgroundColor = .white
        picker.delegate = self
        picker.dataSource = self
        addSubview(picker)
    }

    func installConstraints() {
        blackoutButton.snp.makeConstraints {
            $0.edges.equalTo(self)
        }

        toolbar.snp.makeConstraints {
            $0.bottom.equalTo(self.picker.snp.top)
            $0.height.equalTo(Constants.ToolbarHeight)
            $0.width.equalTo(self)
        }

        picker.snp.makeConstraints {
            $0.width.equalTo(self)
            pickerBottomConstraint = $0.bottom.equalTo(self).constraint
        }
    }

    func show() {
        blackoutButton.alpha = 0.0
        pickerBottomConstraint.update(offset: picker.frame.size.height + Constants.ToolbarHeight)

        layoutIfNeeded()

        UIView.animate(withDuration: Constants.AnimationDuration, animations: {
            self.blackoutButton.alpha = 1.0
            self.pickerBottomConstraint.update(offset: 0.0)

            self.layoutIfNeeded()
            UIAccessibilityPostNotification(UIAccessibilityScreenChangedNotification, self.picker);
        }) 
    }

    func hide() {
        UIView.animate(withDuration: Constants.AnimationDuration, animations: {
            self.blackoutButton.alpha = 0.0
            self.pickerBottomConstraint.update(offset: self.picker.frame.size.height + Constants.ToolbarHeight)

            self.layoutIfNeeded()
        }, completion: { finished in
            self.removeFromSuperview()
        })
    }
}
